### PR TITLE
Remove version overwriting hacks

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
### What was the end-user problem that led to this PR?

The problem was that we are doing some hacks in our version file that are... dangerous. At very least they cause "circular problems" like #6927, but I also have bad feelings about it.

### What was your diagnosis of the problem?

My diagnosis was we're overwriting the version of the currently loaded bundler :grimacing:. This hack is _almost_ unnecessary (see the spec run here with the hack fully removed: https://travis-ci.org/bundler/bundler/builds/509497021. Only three jobs failed, and for old rubies/rubygems/bundler). 

### What is your fix for the problem, implemented in this PR?

My fix is to limit the hack to `bundler`'s spec run, since it's only needed there.

### Why did you choose this fix out of the possible options?

I chose this fix because fully getting the build green with the hack fully removed is a bit of work. I dedicated a bit of time to investigate one of the failures and it all comes down to the priority of loading default gems vs `-I`. So I think something like https://github.com/rubygems/rubygems/pull/1868 should fix it. But that's out of the scope of this PR.

Closes #6927.

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
Ok, so I looked a bit more into this, and it turns out that this "version hack" is only needed for rubygems versions the doesn't provide the `BundlerVersionFinder`. So I added a check for the rubygems version, and removed the hack.

While debugging this, I added a few other changes that are not directly related to this, so I have extracted them to separate PRs (#7136, #7137, #7138, #7139).

@segiddins What do you think about this? I really don't like changing the version of a loaded gem, if we can avoid it :grimacing:...

File: lib/bundler.rb [hunk 1]
```diff
@@ -27,36 +27,36 @@ module Bundler
   autoload :DepProxy,               "bundler/dep_proxy"
   autoload :Deprecate,              "bundler/deprecate"
   autoload :Dsl,                    "bundler/dsl"
-  autoload :EndpointSpecification,  "bundler/endpoint_specification"
+  autoload :EndpointSpecification,  File.expand_path("bundler/endpoint_specification", __dir__)
```
Comment (segiddins) [inline lib/bundler.rb]: this seems... not right to me

File: .travis.yml [hunk 2]
```diff
@@ -13,6 +13,7 @@ branches:
     - trying
     - /.+-dev$/
     - /.+-stable$/
+    - more_relative_code_loading
```
Comment (segiddins) [inline .travis.yml]: revert

File: spec/bundler/shared_helpers_spec.rb [hunk 3]
```diff
@@ -261,15 +261,6 @@
       subject.set_bundle_environment
     end
 
-    it "ignores if bundler_ruby_lib is same as rubylibdir" do
```
Comment (segiddins) [inline spec/bundler/shared_helpers_spec.rb]: why is this spec removed?

File: spec/runtime/load_spec.rb [hunk 4]
```diff
@@ -56,19 +57,6 @@
         Bundler.load
       end.to raise_error(Bundler::GemfileNotFound, /omg\.rb/)
     end
-
-    it "does not find a Gemfile above the testing directory" do
```
Comment (segiddins) [inline spec/runtime/load_spec.rb]: why is this removed?

**Comment by deivid-rodriguez** [comment]
Sorry @segiddins. This is not ready for review. A general idea of whether you think this is a good idea is appreciated though.

File: lib/bundler.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/bundler.rb]: This is just an experiment not meant to stay in this PR, but I actually planned to open a separate PR to migrate our autoloads to require absolute paths. I think it would be a nice follow up to https://github.com/bundler/bundler/pull/7100 since we would make sure that we never use the $LOAD_PATH for requiring our own internal code.

What do you think it is not right?

By the way, it would be prettier if we had this: https://github.com/bundler/bundler/pull/7100 :smiley:.

File: .travis.yml (hunk 2 above)
Comment (deivid-rodriguez) [reply .travis.yml]: WIP code, not ready for review.

File: spec/bundler/shared_helpers_spec.rb (hunk 3 above)
Comment (deivid-rodriguez) [reply spec/bundler/shared_helpers_spec.rb]: This is part of an experiment of the hypothesis that if we never use the $LOAD_PATH, we can revert several patches and workarounds for issues introduced in the past, and we no longer need to special case bundler when manipulating the $LOAD_PATH.

But it's not meant to stay in this PR.

File: spec/runtime/load_spec.rb (hunk 4 above)
Comment (deivid-rodriguez) [reply spec/runtime/load_spec.rb]: This is part of https://github.com/bundler/bundler/pull/7138, can you review that PR instead for the rationale? That one is ready for review :+1:.

**Comment by deivid-rodriguez** [comment]
@segiddins This should be ready for a review now!

File: lib/bundler.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/bundler.rb]: Marking this as solved because I removed the changes from this PR. I'll open a separate discussion for migrating our autoloads to not use the load path.

**Comment by deivid-rodriguez** [comment]
These version overwriting hacks bit me again in #7166, but I think, at least some of them can be fixed for all supported rubygems versions, so I'm back to working on this PR :sweat_smile:. I'll re-request a review when I'm ready :)

**Comment by deivid-rodriguez** [comment]
I'm so happy I finally managed to keep this part of the diff

```diff
diff --git a/lib/bundler/version.rb b/lib/bundler/version.rb
index cfd630095..21eeeb0fd 100644
--- a/lib/bundler/version.rb
+++ b/lib/bundler/version.rb
@@ -1,23 +1,7 @@
 # frozen_string_literal: false
 
 module Bundler
-  # We're doing this because we might write tests that deal
-  # with other versions of bundler and we are unsure how to
-  # handle this better.
-  VERSION = "2.1.0.pre.1".freeze unless defined?(::Bundler::VERSION)
-
-  def self.overwrite_loaded_gem_version
-    begin
-      require "rubygems"
-    rescue LoadError
-      return
-    end
-    return unless bundler_spec = Gem.loaded_specs["bundler"]
-    return if bundler_spec.version == VERSION
-    bundler_spec.version = Bundler::VERSION
-  end
-  private_class_method :overwrite_loaded_gem_version
-  overwrite_loaded_gem_version
+  VERSION = "2.1.0.pre.1".freeze
 
   def self.bundler_major_version
     @bundler_major_version ||= VERSION.split(".").first.to_i
```

while also keeping CI green for all supported rubygems & rubies :tada:

The price to pay is the admittedly ugly style of using absolute paths with autoloads, but I think it's definitely worth it. We could make it prettier if [autoload_relative ever lands into core](https://bugs.ruby-lang.org/issues/15330).

This should fix a bunch of problems with manually installed bundler versions coexisting with bundler versions as default gems :smiley:.

As per vendored dependencies, I created https://github.com/CocoaPods/Molinillo/pull/130 for `Molinillo`, and `fileutils` has already been updated but changes have not yet been incorporated here (I created #7156 to do that separately). The more complicated one would be `thor` because it still supports 1.8.7. I requested that to be dropped in https://github.com/erikhuda/thor/pull/643#issuecomment-490611243. In any case, the activity in the `thor` repo is really low, so I think it's fine to incorporate these changes directly, and propose them upstream once they drop support.

I'd like other to chime in here :smiley:

**Comment by deivid-rodriguez** [comment]
Also, I think this might get green several specs that are currently skipped when running with the "ruby-core" flag.

**Comment by deivid-rodriguez** [comment]
Thanks @indirect!

@bundlerbot r+
